content/recipes/authoring/kata-snippets/full-tests.md

```diff
@@ -5,6 +5,84 @@ prev: /recipes/authoring/kata-snippets/sample-tests/
+  - Reference solution being accessible to users by mistake
+  - Input mutation by the user solution affecting the input passed to the reference solution
+  - Incorrect implementation of the reference solution leading to rejection of valid user solutions
+- **Reference solution should not be revealed to the user.** When an assertion fails or the test suite crashes, some testing frameworks print fragments of source code which caused the failure to the console. It may happen that such printed failure messages or stack traces expose information about the solution which should not be revealed.
+- **Reference solution shouldn't be accessible to user solution.** It should not be possible to implement the user solution as an alias or wrapper around the reference solution. The reference solution should be completely inaccessible outside the submit tests. In particular, it should not be a global function.
+Chek the [reference page for your language](/languages/) to see how to prevent this problem in your tests.
+
+
+## Fixed Tests
+
+Fixed tests are tests with predetermined inputs, and do not change between test runs.
+
+- **Each requirement should have a corresponding fixed test.** Every requirement in the kata description should be explicitly tested with at least one, possibly more, dedicated and properly labeled assertion(s).
+- **Tests should check not only valid inputs but also invalid inputs and edge cases** (if the problem allows them). `null` inputs, empty arrays, empty strings, etc. should all be tested with fixed tests, unless the kata description _explicitly_ states that such inputs need not be considered. Random tests are not a replacement for carefully designed assertions for edge cases.
+- If your test suite uses random tests (and it usually should, see below), then make sure that every test scenario generated there has its equivalent fixed test. Fixed tests are usually easier to debug, so ideally a fixed test should fail before the corresponding scenario is tested with random tests.
+
+
+## Random Tests
+
+Random tests are uncommon in "real life" coding and are somewhat specific to Codewars. They are required by many kata to reject solutions based on input probing, hard-coding, and other workarounds. The goal of random tests is to make the inputs unpredictable so that only proper solutions may pass.
+
+- **Each requirement should have at least one (preferably more) corresponding random test case.** Basically, randomly generated scenarios should correspond to the set of fixed tests. You should not rely on luck and hope that random tests will generate invalid inputs and edge conditions by chance.
+- **Random inputs should be _truly_ (pseudo-)random**, i.e. they should be _dynamically_ generated on every run with the help of suitable library functions such as `Math.random()` in JavaScript or `rand()` in C. See this [obligatory XKCD comic](https://xkcd.com/221/) for how NOT to do random tests.
```

Comment:
on this and the following: the XKCD has to be kept, but the C/C++ note might be too precise for this page (it's language specirfic). Not sure...

Reply:
Yup, I will think about it and get rid of this C thing.

Reply:
- **Use appropriate random utilities available in your language**. Know how to use RNGs, how to randomly generate types of inputs you need, be it numbers in a range, large numbers, strings, big integers, floating point values, collections.  Know how to do random sampling of collections, how to shuffle them, how to avoid duplicates. See this [obligatory XKCD comic](https://xkcd.com/221/) for how NOT to do random tests.